src/structure/locator.rs

```diff
@@ -1,137 +1,128 @@
-use std::convert::From;
+use std::convert::{From, TryInto};
 pub use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};
 
 use speedy::{Context, Readable, Reader, Writable, Writer};
 use serde::{Deserialize, Serialize};
 
 use super::parameter_id::ParameterId;
 
-#[derive(
-  Copy,
-  Clone,
-  Debug,
-  Eq,
-  PartialEq,
-  PartialOrd,
-  Ord,
-  Hash,
-  Readable,
-  Writable,
-  Serialize,
-  Deserialize,
-)]
-pub struct LocatorKind {
-  value: i32,
+pub mod locator_kind {
+  pub const INVALID: i32 = -1;
+  pub const RESERVED: i32 = 0;
+  pub const UDP_V4: i32 = 1;
+  pub const UDP_V6: i32 = 2;
 }
 
-impl LocatorKind {
-  pub const LOCATOR_KIND_INVALID: LocatorKind = LocatorKind { value: -1 };
-  pub const LOCATOR_KIND_RESERVED: LocatorKind = LocatorKind { value: 0 };
-  pub const LOCATOR_KIND_UDP_V4: LocatorKind = LocatorKind { value: 1 };
-  pub const LOCATOR_KIND_UDP_V6: LocatorKind = LocatorKind { value: 2 };
-}
-
-#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize, PartialOrd, Ord)]
-pub struct Locator {
-  pub kind: LocatorKind,
-  pub port: u32,
-  pub address: [u8; 16],
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
+pub enum Locator {
+  Invalid,
+  Reserved,
+  UdpV4(SocketAddr),
+  UdpV6(SocketAddr),
+  Other {
+    discriminant: i32,
+    port: u32,
+    address: [u8; 16],
+  },
```

Comment:
this is perhaps a more idiomatic structure for this type, but this doesn't work since the on-the-wire representation is tightly-coupled to the layout of the type.

you could get around this with the [serde 'from' and 'into' attributes](https://serde.rs/container-attrs.html#into), but this doesn't help you in future if you ever want to support *multiple* serialisation schemes, potentially with different layouts